sentry_sdk/integrations/aws_lambda.py

```diff
@@ -100,7 +101,9 @@ def sentry_handler(event, context, *args, **kwargs):
                     timeout_thread.start()
 
             try:
-                return handler(event, context, *args, **kwargs)
+                transaction = Transaction(op="aws_lambda", name=context.function_name)
```

Comment:
Would prefer something more generic here such as `http.server` or `serverless.function`.

Also this is missing trace propagation entirely. Please use `continue_from_headers` such that we actually get joined traces, not individual transactions. Somewhere in `make_request_event_processor` there should be logic for accessing HTTP headers that we can duplicate. I think it would also be nice if we could instrument functions invoking other functions, but HTTP should be easiest for now

The same comments apply to GCP

Reply:
@untitaker 

I implemented the Transactions using "continue_from_headers" method. I got the same results as before, i.e. we can't capture the internal transactions for 2nd lambda function in case of nested lambda functions.
By "internal transactions" I mean that 2nd lambda function actually calls Dynamo DB & S3 bucket using boto3 library, so internal transactions would involve transactions of API calls(as boto3 functions) to these services.
Also, the trace ID is same for the instrumentation of API call to 2nd lambda function which is happening automatically currently.
I want to know your expectations when you mentioned about the usage of "continue_from_headers" method in the code, as the trace ID is anyways getting passed on, even if we don't use it and even the headers are also getting displayed.

Second thing is that, I want to know about your expectations on Lambda function calling another Lambda function or nested Lambda functions. There are 2 ways as of now to trigger/invoke lambda functions as of now are:-
1. Using boto3 library (POST https://lambda.us-east-2.amazonaws.com/2015-03-31/functions/sentry_dynamo_db/invocations)
2. Using API Gateway (GET https://g8xk68m1n6.execute-api.us-east-2.amazonaws.com/default/sentry_dynamo_db)

In both the above mentioned methods, I've tried to capture traces and both basically involve calling a REST API for 2nd lambda function to be invoked.
So, basically an API call is involved in any case to invoke a lambda function and the invocation traces of the API calls are getting displayed in Sentry dashboard.

Thus, wanted to know if we need to skip the usage for "continue_from_headers" method or still use it even though the results remain the same?
Also, is this change sufficient enough as per the expectations from the code, atleast related to instrumentation of lambda function calling another lambda function?

Reply:
>the invocation traces of the API calls are getting displayed in Sentry dashboard.

yes but if you go down to the trace details, do the two function calls have the same trace ID?

![Screenshot 2020-09-24 at 18 22 35](https://user-images.githubusercontent.com/837573/94172595-f9ed9a00-fe92-11ea-9844-c82be7e6ef39.png)

Imagine a non-serverless (serverful?) stack, such as Django + Celery. There is a Django request handler that spawns a celery task. Those are two transactions, but they should also have the same trace ID. So you can go from the celery task back to the api call and figure out how this data made it into your system.

Imagine two Flask services calling each other via HTTP. All of the transactions of those two services are linked together, because we 1) look at Flask request headers to find sentry-trace header 2) monkeypatch httplib to add sentry-trace to outgoing HTTP requests

Right now this is not really surfaced in the product in an optimal way, but we're working on it.

I don't really understand how function-to-function calls work. If they are not using HTTP, it's probably too hard to implement. But perhaps we can make sure that if the incoming HTTP request contains the sentry-trace header, we honor it instead of creating a new trace ID.

Let me know if that makes sense. I don't think we have to do any of that, if we say it's too hard right now let's just create a new transaction in a new trace. But I just wanted to explain why `continue_from_headers` is different over pure init.

Reply:
Currently function-to-function calls use HTTP and the trace ID is in fact same for both function calls. Look at the trace IDs for both lambda functions and span IDs and parent/child span IDs in the image below.

![trace ID](https://user-images.githubusercontent.com/7957539/94176823-48606000-feb6-11ea-9b27-7b743b4002ed.png)


The implementation that I did right now goes like this:-
API call to first function using Postman   ->   1st Lambda function getting triggered using API Gateway    ->   1st Lambda function (Sentry SDK initialized here) doing API call using requests library in the code to call 2nd Lambda function   ->   2nd Lambda function getting triggered using API Gateway   ->   2nd Lambda function using boto3 library to do operations on Dynamo DB & S3 bucket (which is nothing but API calls)

What I exactly wanted to say was that, since Sentry SDK is initialized in the Lambda function 1, the spans for Lambda function 1 and Lambda function 2 are appearing on the Sentry dashboard as per the the implementation I mentioned above. But, there is no way to capture trace for Dynamo DB and S3 bucket implemented in Lambda function 2 coz Lambda function 2 is not initialized with Sentry SDK. 

Now, If it were the case that we initialized both lambda functions with Sentry SDK and implemented the same way as I mentioned above, then we can capture traces for even Dynamo DB & S3 bucket, and also keeping the span of 2nd lambda function as child of span of 1st lambda function (which is already happening anyways).

Also, I got the idea behind the usage of continue_from_headers and it is easy to implement just wanted to understand reason behind it. Will make the change accordingly as I was able to achieve what you mentioned in the first comment.

The main intention was to match my current understanding of the whole tracing mechanism with yours and to validate if I'm doing the right thing or not. I think it is in alignment now and I can push the final changes soon.